compiler/stable_mir/src/mir/body.rs

```diff
@@ -1007,13 +1007,7 @@ pub enum NullOp {
     /// Returns the offset of a field.
     OffsetOf(Vec<(VariantIdx, FieldIdx)>),
     /// cfg!(debug_assertions), but at codegen time
-    UbCheck(UbKind),
-}
-
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub enum UbKind {
-    LanguageUb,
-    LibraryUb,
+    UbChecks,
```

Comment:
I see no choice but to change smir here; the things it exposed was never meant to be very stable to begin with and I don't think we want to keep them.

Reply:
I've just been breaking smir every time I change MIR. Is there any guidance or enforcement that we don't do that?

Reply:
No idea. SMIR people get pinged when this happens so I guess they'll complain when it gets too much for them. ;)

Reply:
It would be nice to avoid these breakages. If something is unstable, we recommend to not expose them, or expose as an Opaque type.

See Coverage statement as an example.

Reply:
In this specific case, we could just always use the same UbKind value, and mark UbValue as deprecated.

Reply:
There's no UbKind value we can use that reflects the correct semantics, though.

This NullOp is an intrinsic reflected into the MIR, and our intrinsics change all the time. We don't want to be constrained by stability here, that would hinder rustc development.

Reply:
You could turn this into Opaque. So for users that care about this value will need to find out a new way to retrieve it (or request it via `project-stable-mir`, while those that don't care about it, won't be affected.

BTW, I don't know the specifics of every MIR change, these are suggestions. StableMIR is not meant to be immutable, or constraint the evolution of rustc. The goal is to bring a bit more predictability and awareness to APIs that are consumed by other tools.

You as the author of this PR, together with the reviewer, are the best people to know whether a breaking change is worth it. If it changes the semantics of the program, then do go ahead and push it as is. If there is a temporary workaround that can be used to give users some warning of a coming breaking change or minimize the breaking changes, it's even better.

Thank you for bringing this up! We are still learning what's the best way to move forward, and appreciate the patience.

Reply:
Maybe the smir enums should have `Other` variants that can be used to indicate "not currently supported" / "opaque" or so. Then we could map `NullOp::UbChecks` to `NullOp::Other` for now?

But honestly I don't want to figure out how to best do that as part of this PR.

Reply:
If we leave compatibility shims in stable MIR (not that I think we should do that in this PR), is there a plan for how to remove them eventually?

Reply:
I have a question that is not necessarily related to SMIR, which is about this sentence:

> This NullOp type is an intrinsic reflected into the MIR

Why not use `StatementKind::Intrinsic` to model this?

Reply:
StableMIR is not yet versioned, but once it is, we would likely go over things that has been marked as deprecated and remove them when a new major is about to be released.

Reply:
@RalfJung I think it is fine to remove it now.

I do have a question about how to evaluate this operator though. Is the idea here that we are assessing the value of `cfg!(debug_assertions)` for the crate being compiled? So if the MIR comes from a dependency that was compiled with `debug_assertions` off, but the current crate has the `debug_assertions` on, this rvalue will be evaluated to `true`. Is that correct?

 Is it possible to document the answer here?

Reply:
Yes, exactly. This is what the existing documentation means when it says "the value of debug_assertions at monomorphization time".

Reply:
I have also extended the intrinsic's docs to make this more clear.